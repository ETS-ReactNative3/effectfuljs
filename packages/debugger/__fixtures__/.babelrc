{
  "passPerPreset": true,
  "presets": [
    {
      "plugins": [
        [
          "@babel/transform-modules-commonjs",
          { "noInterop": true, "loose": true }
        ]
      ]
    },
    { "plugins": [["../transform",{importRT:"../main"}]] }
  ]
}
