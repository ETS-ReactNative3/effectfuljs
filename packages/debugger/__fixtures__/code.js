const { mod } = require("mod");

let v0 = 10;

function C(a) {
  let v1 = 10;
  try {
    const v2 = 11;
    var v3;
    mod.proc1(mod.fn1(a), v0, v1, v2, v3);
    mod.proc2();
    var v4 = 100;
  } catch (e) {
    mod.proc3(e, v4);
  }
}

class Cl {
  constructor() {
    var v4;
    not_traced(v0, v4);
  }
  get prop1() {
    not_traced();
  }
  set prop1(value) {
    not_traced();
  }
  async funcAsync() {
    not_traced(v0);
  }
  *func() {
    not_traced(() => traced(1) + traced(2));
  }
}

exports.fn2 = function fn1() {
  const v5 = 10;
  return mod.some({ a: mod.f2(), b: [new C(v0, v5)] });
};

const obj1 = {
  objFn(a) {
    var k = new Cl().fn(a, v0);
  }
};
