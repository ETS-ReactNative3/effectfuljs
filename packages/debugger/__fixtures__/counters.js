import * as Kit from "../kit";

export function getCounter() {
  let i = 0;
  return function counter() {
    return i++;
  };
}

export function runCounters() {
  const c = getCounter();
  console.log("C1:", c());
  console.log("C2:", c());
  const state = Kit.capture();
  console.log("C3:", c());
  console.log("C4:", c());
  return [c(), state];
}
