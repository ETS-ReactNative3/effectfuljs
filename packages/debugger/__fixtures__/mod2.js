let cnt = 0;

export function f1(i) {
  cnt++;
  console.log(`f1:${i}@${cnt}`);
  return cnt;
}

export function f2(i) {
  cnt++;
  console.log(`f2:${i}@${cnt}`);
}
