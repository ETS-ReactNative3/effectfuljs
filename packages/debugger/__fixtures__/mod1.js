import * as M2 from "./mod2";

let j = 0;

function run(i = 1) {
  j += i;
  M2.f2(M2.f1(i++), i);
  j += i;
  console.log(`i:${i},j:${j}`);
}

run();
run();
