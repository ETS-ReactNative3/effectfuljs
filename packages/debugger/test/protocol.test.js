const config = require("../config").default;
const path = require("path");
const DEBUG = false;
config.debuggerDebug = DEBUG;
config.verbose = DEBUG;
config.redirConsole = !DEBUG;
config.localConsole = DEBUG;
config.srcRoot = path.resolve(path.join(__dirname, ".."));
config.localConsole = false;
const Kit = require("./protocol");
const D = require("../backends/vscode");
const { context } = require("../state");

describe("VSCode protocol handlers", function() {
  afterEach(Kit.teardown);
  test("step in", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepInPlain"),
      { stopOnEntry: true },
      "stepIn",
      done
    );
  });
  test("`step in` for CommonJS require", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepInCJS"),
      { stopOnEntry: true },
      "stepIn",
      done
    );
  });
  test("`step in` for ESM imports", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepInESM"),
      { stopOnEntry: true },
      "stepIn",
      done
    );
  });
  test("`step in` in async functions", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepInAsync"),
      { stopOnEntry: true },
      "stepIn",
      done
    );
  });
  test("`step in` in generators", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepInGenerator"),
      { stopOnEntry: true },
      "stepIn",
      done
    );
  });
  test("`step in` in async generators", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepInAsyncGenerator"),
      { stopOnEntry: true },
      "stepIn",
      done
    );
  });
  test("`step over` in plain functions", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepOverPlain"),
      { stopOnEntry: true },
      "next",
      done
    );
  });
  test("`step over` for CommonJS require", function(done) {
    Kit.traverse(require("./__fixtures__/stepOverCJS"), {}, "next", done);
  });
  test("`step over` for ESM imports", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepOverESM"),
      { stopOnEntry: true },
      "next",
      done
    );
  });
  test("`step over` in async functions", function(done) {
    Kit.traverse(require("./__fixtures__/stepOverAsync"), {}, "next", done);
  });
  test("`step over` in generator functions", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepOverGenerator"),
      { stopOnEntry: true },
      "next",
      done
    );
  });
  test("`step over` in async generator functions", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepOverAsyncGenerator"),
      {},
      "next",
      done
    );
  });
  test("`step out` in plain functions", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepOutPlain"),
      { stopOnEntry: true },
      "stepOut",
      done
    );
  });
  test("`step out` for CommonJS require", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepOutCJS"),
      { stopOnEntry: true },
      "stepOut",
      done
    );
  });
  test("`step out` for ESM imports", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepOutESM"),
      { stopOnEntry: true },
      "stepOut",
      done
    );
  });
  test("`step out` in async functions", function(done) {
    Kit.traverse(require("./__fixtures__/stepOutAsync"), {}, "stepOut", done);
  });
  test("`step out` in generator functions", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepOutGenerator"),
      {},
      "stepOut",
      done
    );
  });
  test("`step out` in async generator functions", function(done) {
    Kit.traverse(
      require("./__fixtures__/stepOutAsyncGenerator"),
      {},
      "stepOut",
      done
    );
  });
  test("breakpoints", function(done) {
    const PROGRAM = path.resolve(
      config.srcRoot,
      require.resolve("./__fixtures__/breakpoints")
    );
    const SUB_MODULE = path.resolve(
      config.srcRoot,
      require.resolve("./__fixtures__/breakpointSubMod")
    );
    const bps = [
      { line: 5 },
      { line: 8, expLine: 11 },
      { line: 12 },
      { line: 4, expLine: 15 }
    ];
    Kit.awaitEvent("breakpoint", function(event) {
      const bp = bps.find(i => i.id === event.brekpoint.id);
      if (bp) {
        expect(bp.verified).toBe(true);
        expect(bp.line).toBe(bp.expLine || bp.line);
        bp.verified = true;
      }
      return false;
    });
    Kit.traverse(
      /* requied with `expEagerModuleExport: false`:
          (context.moduleId = require.resolve("./__fixtures__/breakpoints")),*/
      require("./__fixtures__/breakpoints"),
      {
        breakpoints: [
          {
            source: {
              path: PROGRAM
            },
            breakpoints: [
              {
                line: 3
              }
            ]
          },
          {
            source: {
              path: SUB_MODULE
            },
            breakpoints: [
              {
                line: 2
              }
            ]
          }
        ],
        onStop(cb, frame) {
          if (frame.line === 3) {
            Kit.command(
              "childSetBreakpoints",
              {
                breakpoints: bps.map(i => ({ line: i.line })),
                source: { path: PROGRAM }
              },
              function(res) {
                res.breakpoints.forEach(function(i, x) {
                  const bp = bps[x];
                  bp.id = x.id;
                  bp.verified = true;
                  if ((bp.verified = i.verified)) {
                    expect(bp.expLine || bp.line).toBe(i.line);
                  }
                });
                cb();
              }
            );
          } else cb();
        }
      },
      "continue",
      function() {
        expect(bps.every(i => i.verified)).toBe(true);
        done();
      }
    );
  });
  test("conditional breakpoints", function(done) {
    const PROGRAM = path.resolve(
      config.srcRoot,
      require.resolve("./__fixtures__/conditionalBreakpoints")
    );
    Kit.traverse(
      require("./__fixtures__/conditionalBreakpoints"),
      {
        breakpoints: [
          {
            source: {
              path: PROGRAM
            },
            breakpoints: [
              {
                line: 3,
                condition: "i === 3"
              }
            ]
          }
        ]
      },
      "continue",
      done
    );
  });
  test("watch expression", function(done) {
    const vals = [];
    Kit.traverse(
      require("./__fixtures__/watchExpr"),
      {
        stopOnEntry: true,
        onStop(cb) {
          Kit.command(
            "evaluate",
            {
              context: "watch",
              expression: "`i:${i},j:${j}`",
              frameId: 0
            },
            function(resp) {
              vals.push(resp.result);
              Kit.command(
                "evaluate",
                {
                  context: "watch",
                  expression: "{i:i,j:j}",
                  frameId: 0
                },
                function(resp) {
                  if (resp.variablesReference)
                    Kit.varValue(resp.variablesReference, 2, function(value) {
                      vals.push(value);
                      cb();
                    });
                  else {
                    vals.push(resp.result);
                    cb();
                  }
                }
              );
            }
          );
        }
      },
      "stepIn",
      function() {
        try {
          expect(vals).toMatchSnapshot();
          done();
        } catch (e) {
          done(e);
        }
      }
    );
  });
  test("logpoint", function(done) {
    const vals = [];
    Kit.traverse(
      require("./__fixtures__/logPoints"),
      {
        breakpoints: [
          {
            source: {
              path: path.resolve(
                config.srcRoot,
                require.resolve("./__fixtures__/logPoints")
              )
            },
            breakpoints: [
              {
                line: 3,
                logMessage: "i:{i}, j:{j}"
              }
            ]
          }
        ]
      },
      "continue",
      done
    );
  });
  test("hit conditions", function(done) {
    const vals = [];
    Kit.traverse(
      require("./__fixtures__/hitCond"),
      {
        breakpoints: [
          {
            source: {
              path: path.resolve(
                config.srcRoot,
                require.resolve("./__fixtures__/hitCond")
              )
            },
            breakpoints: [
              {
                line: 3,
                hitCondition: "HIT_AT-2"
              }
            ]
          }
        ]
      },
      "continue",
      done
    );
  });
  // different paths on server TODO:
  test.skip("all exception breakpoints", function(done) {
    Kit.traverse(
      require("./__fixtures__/exceptionBreakpoints"),
      {
        exceptions: {
          filters: ["all"]
        }
      },
      "continue",
      done
    );
  });
  // different paths on server TODO:
  test.skip("uncaught exception breakpoints", function(done) {
    Kit.traverse(
      require("./__fixtures__/exceptionBreakpointsUncaught"),
      {
        stopOnEntry: true,
        onStop(cb, frame) {
          if (frame.line === 1)
            Kit.command(
              "childSetExceptionBreakpoints",
              { filters: ["uncaught"] },
              cb
            );
          else cb();
        }
      },
      "continue",
      done
    );
  });
  test("stepping in `eval`", function(done) {
    let num = 0;
    Kit.traverse(
      require("./__fixtures__/evalDir"),
      {
        onStop(cb, frame) {
          num++;
          if (num === 1) {
            Kit.command(
              "childSetBreakpoints",
              {
                breakpoints: [
                  {
                    line: 4,
                    condition: "i === 3"
                  }
                ],
                source: { sourceReference: frame.file }
              },
              function(res) {
                const {
                  breakpoints: [bp]
                } = res;
                expect(bp.line).toBe(4);
                expect(bp.verified).toBe(true);
                cb();
              }
            );
          } else if (num === 2) {
            Kit.command(
              "evaluate",
              {
                context: "watch",
                expression: "`i:${i},j:${j}`",
                frameId: 0
              },
              function(resp) {
                expect(resp.result).toBe("i:3,j:1");
                cb();
              }
            );
          } else cb();
        }
      },
      "continue",
      done
    );
  });
  test("stepping in indirect `eval`", function(done) {
    let num = 0;
    Kit.traverse(
      require("./__fixtures__/evalIndir"),
      {
        onStop(cb, frame) {
          num++;
          if (num === 1) {
            Kit.command(
              "childSetBreakpoints",
              {
                breakpoints: [
                  {
                    line: 4,
                    condition: "i === 3"
                  }
                ],
                source: { sourceReference: frame.file }
              },
              function(res) {
                const {
                  breakpoints: [bp]
                } = res;
                expect(bp.line).toBe(4);
                expect(bp.verified).toBe(true);
                cb();
              }
            );
          } else if (num === 2) {
            expect(frame.line).toBe(4);
            Kit.command(
              "evaluate",
              {
                context: "watch",
                expression: "`i:${i},j:${j}`",
                frameId: 0
              },
              function(resp) {
                expect(resp.result).toBe("i:3,j:2");
                cb();
              }
            );
          } else cb();
        }
      },
      "continue",
      done
    );
  });
  test("stepping in Function constructor", function(done) {
    Kit.traverse(require("./__fixtures__/funconstr"), {}, "stepOut", done);
  });
  // this breaks Promises (leaves them in inconsistent state)
  // so works only if run last (TODO: adapt test)
  test.skip("pause", function(done) {
    let pauseDone = false;
    Kit.launch(require("./__fixtures__/pause"), {}, function() {
      Kit.awaitEvent("stopped", function(args) {
        expect(args.reason).toBe("step");
        Kit.command("stackTrace", {}, function(stack) {
          try {
            const frame = stack.stackFrames[0];
            expect(pauseDone).toBe(true);
            expect(frame.name).toBe("main");
            expect(frame.line).toBe(3);
            done();
          } catch (e) {
            done(e);
          }
        });
      });
      setTimeout(function() {
        Kit.command("pause", {}, function() {
          pauseDone = true;
        });
      }, 100);
    });
  });
});
