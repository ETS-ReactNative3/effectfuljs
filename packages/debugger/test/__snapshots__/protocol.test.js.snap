// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VSCode protocol handlers breakpoints 1`] = `
Object {
  "stderr": "",
  "stdout": "",
  "trace": Array [
    Object {
      "reason": "breakpoint",
      "stack": Array [
        Object {
          "column": 1,
          "file": "breakpoints.js",
          "line": 3,
          "name": "f",
          "scopes": Array [
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "1",
            },
          ],
        },
      ],
    },
    Object {
      "reason": "breakpoint",
      "stack": Array [
        Object {
          "column": 1,
          "file": "breakpoints.js",
          "line": 15,
          "name": "f",
          "scopes": Array [
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "2",
            },
          ],
        },
      ],
    },
    Object {
      "reason": "breakpoint",
      "stack": Array [
        Object {
          "column": 20,
          "file": "breakpoints.js",
          "line": 5,
          "name": "func1",
          "scopes": Array [
            Object {
              "[name]": "func1",
            },
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "3",
            },
          ],
        },
        Object {
          "column": 1,
          "file": "breakpoints.js",
          "line": 16,
          "name": "f",
          "scopes": Array [
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "3",
            },
          ],
        },
      ],
    },
    Object {
      "reason": "breakpoint",
      "stack": Array [
        Object {
          "column": 1,
          "file": "breakpoints.js",
          "line": 11,
          "name": "func2",
          "scopes": Array [
            Object {
              "[name]": "func2",
              "func2": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func2",
                "prototype": Object {
                  "constructor": "func2@breakpoints.js:5:20-13:1",
                },
              },
            },
            Object {
              "[name]": "func1",
            },
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "3",
            },
          ],
        },
        Object {
          "column": 20,
          "file": "breakpoints.js",
          "line": 5,
          "name": "func1",
          "scopes": Array [
            Object {
              "[name]": "func1",
            },
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "3",
            },
          ],
        },
        Object {
          "column": 1,
          "file": "breakpoints.js",
          "line": 16,
          "name": "f",
          "scopes": Array [
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "3",
            },
          ],
        },
      ],
    },
    Object {
      "reason": "breakpoint",
      "stack": Array [
        Object {
          "column": 12,
          "file": "breakpoints.js",
          "line": 12,
          "name": "func2",
          "scopes": Array [
            Object {
              "[name]": "func2",
              "func2": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func2",
                "prototype": Object {
                  "constructor": "func2@breakpoints.js:5:20-13:1",
                },
              },
            },
            Object {
              "[name]": "func1",
            },
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "4",
            },
          ],
        },
        Object {
          "column": 20,
          "file": "breakpoints.js",
          "line": 5,
          "name": "func1",
          "scopes": Array [
            Object {
              "[name]": "func1",
            },
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "4",
            },
          ],
        },
        Object {
          "column": 1,
          "file": "breakpoints.js",
          "line": 16,
          "name": "f",
          "scopes": Array [
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "4",
            },
          ],
        },
      ],
    },
    Object {
      "reason": "breakpoint",
      "stack": Array [
        Object {
          "column": 12,
          "file": "breakpoints.js",
          "line": 12,
          "name": "func2",
          "scopes": Array [
            Object {
              "[name]": "func2",
              "func2": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func2",
                "prototype": Object {
                  "constructor": "func2@breakpoints.js:5:20-13:1",
                },
              },
            },
            Object {
              "[name]": "func1",
            },
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "5",
            },
          ],
        },
        Object {
          "column": 20,
          "file": "breakpoints.js",
          "line": 5,
          "name": "func1",
          "scopes": Array [
            Object {
              "[name]": "func1",
            },
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "5",
            },
          ],
        },
        Object {
          "column": 1,
          "file": "breakpoints.js",
          "line": 16,
          "name": "f",
          "scopes": Array [
            Object {
              "[name]": "f",
              "func1": Object {
                "apply": Object {
                  "length": "0",
                  "name": "defaultApply",
                  "prototype": "[object Object]",
                },
                "call": Object {
                  "length": "0",
                  "name": "defaultCall",
                  "prototype": "[object Object]",
                },
                "length": "0",
                "name": "func1",
                "prototype": Object {
                  "constructor": "func1@breakpoints.js:5:0-14:1",
                },
              },
              "i": "5",
            },
          ],
        },
      ],
    },
  ],
}
`;
